File: euler/14/14.py
#!/usr/bin/env python

def do_seq(x):
    count = 0
    while True:
        count +=1
        if x==1:
            return count
        if x%2==0:
            x=x/2
        else:
            x=(3*x)+1

# Find longest sequence starting under 1,000,000
mchain = 0
mstart = 2
for x in range(2,999999):
    chain = do_seq(x)
    if (chain>mchain):
        mchain=chain
        mstart=x
    if (x%10000==0):
        print "Currently on starting value %d" % (x)

print "Maximum chain starts on %d and has length %d" % (mstart,mchain)

File: euler/14/14.f90
program problem14

integer (kind=8) :: do_seq,x,mchain,mstart,chain,junk
mchain=0
mstart=0

do x=2,999999
    chain=do_seq(x)
    if (chain>mchain) then
        mchain=chain
        mstart=x
    endif
    if (mod(x,10000).eq.0) then
        print*,"Currently on x = ",x,"max length is",mchain,"with start",mstart
    endif
enddo

print*,"Max chain length ",mchain,"starts at ",mstart

end program

function do_seq(n)
    integer (kind=8) :: do_seq,x,n,c,j
    x=n
    c=0
    do while (x.ne.1)
        c=c+1
        if (mod(x,2).eq.0) then
            x=x/2
        else
            x=(3*x)+1
        endif
        if (c>10000) then
            print*,"Seems to be stuck on ",n
            read*,j
        endif
    enddo
    do_seq=c
    return 
end function
